### PR TITLE
MultiInferenceMessage & MultiResponseMessage share a new base class

### LINKED ISSUE #311: [BUG] ResponseMemoryProbs not used by MultiResponseProbsMessage
**Describe the bug**
`ResponseMemoryProbs` not used by `MultiResponseProbsMessage`

**Steps/Code to reproduce bug**
From reading the code its clear that `MultiResponseProbsMessage` was intended to use `ResponseMemoryProbs`, but is instead using the parent class `ResponseMemory`

Other parts of the code-base do use `ResponseMemoryProbs`.

We should do either

1. Remove `ResponseMemoryProbs` entirely, potentially updating either `ResponseMemory` or `TensorMemory` to optionally enforce the inclusion of specific tensors.
2. Actually use `ResponseMemoryProbs`

### LINKED ISSUE #313: [FEA] MultiInferenceMessage & MultiResponseMessage should have a new common base class
**Is your feature request related to a problem? Please describe.**
Both classes inherit from `MultiMessage` and add the same three new attributes:
```
std::shared_ptr<InferenceMemory | ResponseMemory> memory;
    std::size_t offset{0};
    std::size_t count{0};

```

Both `InferenceMemory` and `ResponseMemory` have the same common base `TensorMemory`

This would allow both message classes to share some common code for interacting with a named collection of tensors

**Describe the solution you'd like**
New base class named ? MultiTensorMessage ?

### PR DESCRIPTION
**Pull request by dagardner-nv**
* `MultiInferenceMessage` & `MultiResponseMessage` now inherit a new class `MultiTensorMessage` reducing the duplicated code in each
* `MultiResponseProbsMessage` class now uses `ResponseMemoryProbs`
* Add missing docstrings
* Common CSV & Json reading code moved to deserializers.[hpp|cpp]

fixes #313 
fixes #311

### CODE REVIEW

File: morpheus/_lib/include/morpheus/messages/memory/inference_memory.hpp [hunk 1]
```diff
@@ -43,7 +45,6 @@ class InferenceMemory : public TensorMemory
 };
 
 /****** InferenceMemoryInterfaceProxy *************************/
-#pragma GCC visibility push(default)
```
Comment (dagardner-nv) [inline morpheus/_lib/include/morpheus/messages/memory/inference_memory.hpp:46]: Moves the visibility pragma to cover the `InferenceMemory` class

File: morpheus/_lib/include/morpheus/messages/multi_response.hpp [hunk 2]
```diff
@@ -47,39 +48,42 @@ class MultiResponseMessage : public DerivedMultiMessage<MultiResponseMessage, Mu
...
+     * @brief Returns the output tensor with the given name. Will halt on a fatal error if the tensor does not exist.
+     *
+     * @param name
+     * @return const TensorObject
      */
-    const TensorObject get_output(const std::string &name) const;
+    const TensorObject get_output(const std::string &name) const
+    {
+        return get_tensor(name);
+    };
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/messages/multi_response.hpp:63]: Why are these not implemented in a .cpp

File: morpheus/_lib/src/stages/file_source.cpp [hunk 3]
```diff
@@ -114,54 +92,6 @@ FileSourceStage::subscriber_fn_t FileSourceStage::build()
...
-cudf::io::table_with_metadata FileSourceStage::load_table()
-{
-    auto file_path = std::filesystem::path(m_filename);
-
-    if (file_path.extension() == ".json" || file_path.extension() == ".jsonlines")
-    {
-        // First, load the file into json
-        auto options = cudf::io::json_reader_options::builder(cudf::io::source_info{m_filename}).lines(true);
-
-        auto tbl = cudf::io::read_json(options.build());
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/file_source.cpp:126]: This is a good catch. Was going to suggest you do this as well but you beat me to it.

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Only changes I would like to see is some of the implementations moved to a CPP files. I made 1 comment but there are many occurances

File: morpheus/_lib/include/morpheus/messages/multi_inference.hpp [hunk 4]
```diff
@@ -47,34 +48,29 @@ class MultiInferenceMessage : public DerivedMultiMessage<MultiInferenceMessage,
                           std::size_t mess_count,
                           std::shared_ptr<morpheus::InferenceMemory> memory,
                           std::size_t offset,
-                          std::size_t count);
-
-    std::shared_ptr<morpheus::InferenceMemory> memory;
-    std::size_t offset{0};
-    std::size_t count{0};
+                          std::size_t count) :
+      DerivedMultiMessage(meta, mess_offset, mess_count, memory, offset, count){};
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/messages/multi_inference.hpp:52]: This should be in the cpp file.

File: morpheus/_lib/include/morpheus/messages/multi_response.hpp [hunk 5]
```diff
@@ -47,39 +48,33 @@ class MultiResponseMessage : public DerivedMultiMessage<MultiResponseMessage, Mu
...
-    std::shared_ptr<ResponseMemory> memory;
-    std::size_t offset{0};
-    std::size_t count{0};
-
-    /**
-     * TODO(Documentation)
-     */
-    TensorObject get_output(const std::string &name);
+                         std::size_t count) :
+      DerivedMultiMessage(meta, mess_offset, mess_count, memory, offset, count){};
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/messages/multi_response.hpp:52]: Move to cpp file

File: morpheus/_lib/include/morpheus/messages/multi_tensor.hpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+ * Contains a pointer to an instance of TensorMemory along with an
+ * offset & count to those tensors.
+ *
+ * mess_offset & mess_count refer to the range of records in meta.
+ * offset & count refer to the range of records in TensorMemory.
+ *
+ * While TensorMemory can contain multiple tensors, it is a requirement that
+ * they are all of the same length and that element N in each tensor refers
+ * to the same record.
+ */
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/messages/multi_tensor.hpp:44]: This should be moved after the `#pragma` so it is touching the class

File: morpheus/_lib/src/messages/multi_inference.cpp [hunk 7]
```diff
@@ -131,7 +66,7 @@ std::shared_ptr<MultiInferenceMessage> MultiInferenceMessageInterfaceProxy::init
 
 std::shared_ptr<morpheus::InferenceMemory> MultiInferenceMessageInterfaceProxy::memory(MultiInferenceMessage &self)
 {
-    return self.memory;
+    return DCHECK_NOTNULL(std::dynamic_pointer_cast<morpheus::InferenceMemory>(self.memory));
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_inference.cpp:69]: Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.

File: morpheus/_lib/src/messages/multi_inference_fil.cpp [hunk 8]
```diff
@@ -73,7 +73,7 @@ std::shared_ptr<MultiInferenceFILMessage> MultiInferenceFILMessageInterfaceProxy
 std::shared_ptr<morpheus::InferenceMemory> MultiInferenceFILMessageInterfaceProxy::memory(
     MultiInferenceFILMessage &self)
 {
-    return self.memory;
+    return DCHECK_NOTNULL(std::dynamic_pointer_cast<morpheus::InferenceMemory>(self.memory));
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_inference_fil.cpp:76]: Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.

File: morpheus/_lib/src/messages/multi_inference_nlp.cpp [hunk 9]
```diff
@@ -86,7 +86,7 @@ std::shared_ptr<MultiInferenceNLPMessage> MultiInferenceNLPMessageInterfaceProxy
 std::shared_ptr<morpheus::InferenceMemory> MultiInferenceNLPMessageInterfaceProxy::memory(
     MultiInferenceNLPMessage &self)
 {
-    return self.memory;
+    return DCHECK_NOTNULL(std::dynamic_pointer_cast<morpheus::InferenceMemory>(self.memory));
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_inference_nlp.cpp:89]: Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.

File: morpheus/_lib/src/messages/multi_response.cpp [hunk 10]
```diff
@@ -141,7 +66,7 @@ std::shared_ptr<MultiResponseMessage> MultiResponseMessageInterfaceProxy::init(s
 
 std::shared_ptr<morpheus::ResponseMemory> MultiResponseMessageInterfaceProxy::memory(MultiResponseMessage &self)
 {
-    return self.memory;
+    return DCHECK_NOTNULL(std::dynamic_pointer_cast<morpheus::ResponseMemory>(self.memory));
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_response.cpp:69]: Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.

File: morpheus/_lib/src/messages/multi_response_probs.cpp [hunk 11]
```diff
@@ -56,18 +56,18 @@ std::shared_ptr<MultiResponseProbsMessage> MultiResponseProbsMessageInterfacePro
...
     return std::make_shared<MultiResponseProbsMessage>(
         std::move(meta), mess_offset, mess_count, std::move(memory), offset, count);
 }
 
-std::shared_ptr<morpheus::ResponseMemory> MultiResponseProbsMessageInterfaceProxy::memory(
+std::shared_ptr<morpheus::ResponseMemoryProbs> MultiResponseProbsMessageInterfaceProxy::memory(
     MultiResponseProbsMessage &self)
 {
-    return self.memory;
+    return DCHECK_NOTNULL(std::dynamic_pointer_cast<morpheus::ResponseMemoryProbs>(self.memory));
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_response_probs.cpp:70]: Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.

File: morpheus/_lib/src/messages/multi_tensor.cpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+{
+    CHECK(this->memory->has_tensor(name)) << "Cound not find tensor: " << name;
+
+    // check if we are getting the entire input
+    if (this->offset == 0 && this->count == this->memory->count)
+    {
+        return this->memory->tensors[name];
+    }
+
+    // TODO(MDD): This really needs to return the slice of the tensor
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_tensor.cpp:63]: I think this comment is no longer needed.

File: morpheus/_lib/src/messages/multi_tensor.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+    // Pass onto the base
+    DerivedMultiMessage::get_slice_impl(new_message, start, stop);
+}
+
+void MultiTensorMessage::copy_ranges_impl(std::shared_ptr<MultiMessage> new_message,
+                                          const std::vector<std::pair<size_t, size_t>> &ranges,
+                                          size_t num_selected_rows) const
+{
+    auto copied_message = DCHECK_NOTNULL(std::dynamic_pointer_cast<MultiTensorMessage>(new_message));
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_tensor.cpp:105]: Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.

File: morpheus/_lib/src/messages/multi_tensor.cpp [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+    // Set the value to use assignment
+    slice = value;
+}
+
+void MultiTensorMessage::get_slice_impl(std::shared_ptr<MultiMessage> new_message,
+                                        std::size_t start,
+                                        std::size_t stop) const
+{
+    auto sliced_message = DCHECK_NOTNULL(std::dynamic_pointer_cast<MultiTensorMessage>(new_message));
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_tensor.cpp:81]: Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.

File: morpheus/_lib/src/stages/triton_inference.cpp [hunk 15]
```diff
@@ -113,7 +113,7 @@ InferenceClientStage::subscribe_fn_t InferenceClientStage::build_operator()
                 // When our tensor lengths are longer than our dataframe we will need to use the seq_ids
                 // array to lookup how the values should map back into the dataframe
                 const bool needs_seq_ids = x->mess_count != x->count;
-                auto reponse_memory      = std::make_shared<ResponseMemory>(x->mess_count);
+                ResponseMemory reponse_memory(x->mess_count);
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/triton_inference.cpp:116]: Creating a value type of `ResponseMemory` just to move it into a `shared_ptr<ResponseMemoryProbs>` is awkward. Instead,

1. Dont create a `ReponseMemory` object at all. Just create a `std::map<std::string, TensorObject>` and store all of the tensors in there.
2. Create a new constructor for `ResponseMemoryProbs` that accepts a `tensor_map_t` object
   3. Similar to the move constructor overload in `ResponseMemoryProbs`, check for the tensor named `"probs"`
4. After the for loop that creates all of the `TensorObject`, create a `ResponseMemoryProbs` using the new constructor overload

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Couple of redundant changes regarding static pointer casts. If that doesnt work out, disregard them all.

File: morpheus/_lib/src/stages/triton_inference.cpp (hunk 15 above)
Comment (dagardner-nv) [reply morpheus/_lib/src/stages/triton_inference.cpp:116]: Agreed that makes it much better

File: morpheus/_lib/src/messages/memory/response_memory_probs.cpp [hunk 16]
```diff
@@ -36,10 +36,10 @@ ResponseMemoryProbs::ResponseMemoryProbs(size_t count, TensorObject probs) : Res
     this->tensors["probs"] = std::move(probs);
 }
 
-ResponseMemoryProbs::ResponseMemoryProbs(ResponseMemory &&other) : ResponseMemory(other.count, std::move(other.tensors))
+ResponseMemoryProbs::ResponseMemoryProbs(size_t count, tensor_map_t &&tensors) :
+  ResponseMemory(count, std::move(tensors))
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/memory/response_memory_probs.cpp:40]:
```suggestion
ResponseMemory(count, std::move(tensors))
```

File: morpheus/_lib/src/messages/multi_response.cpp [hunk 17]
```diff
@@ -66,7 +74,8 @@ std::shared_ptr<MultiResponseMessage> MultiResponseMessageInterfaceProxy::init(s
 
 std::shared_ptr<morpheus::ResponseMemory> MultiResponseMessageInterfaceProxy::memory(MultiResponseMessage &self)
 {
-    return DCHECK_NOTNULL(std::dynamic_pointer_cast<morpheus::ResponseMemory>(self.memory));
+    DCHECK(std::dynamic_pointer_cast<morpheus::ResponseMemory>(self.memory) != nullptr);
+    return std::static_pointer_cast<morpheus::ResponseMemory>(self.memory);
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_response.cpp:78]:
```suggestion
    DCHECK(std::dynamic_pointer_cast<morpheus::ResponseMemory>(self.memory) != nullptr);

    return std::static_pointer_cast<morpheus::ResponseMemory>(self.memory);
```
Match the spacing of the others

File: morpheus/_lib/src/messages/multi_response_probs.cpp [hunk 18]
```diff
@@ -67,7 +67,8 @@ std::shared_ptr<MultiResponseProbsMessage> MultiResponseProbsMessageInterfacePro
 std::shared_ptr<morpheus::ResponseMemoryProbs> MultiResponseProbsMessageInterfaceProxy::memory(
     MultiResponseProbsMessage &self)
 {
-    return DCHECK_NOTNULL(std::dynamic_pointer_cast<morpheus::ResponseMemoryProbs>(self.memory));
+    DCHECK(std::dynamic_pointer_cast<morpheus::ResponseMemoryProbs>(self.memory) != nullptr);
+    return std::static_pointer_cast<morpheus::ResponseMemoryProbs>(self.memory);
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/messages/multi_response_probs.cpp:71]:
```suggestion
    DCHECK(std::dynamic_pointer_cast<morpheus::ResponseMemoryProbs>(self.memory) != nullptr);

    return std::static_pointer_cast<morpheus::ResponseMemoryProbs>(self.memory);
```
Match spacing

**Review by mdemoret-nv — APPROVED** [review]
Looks good.

File: morpheus/_lib/src/messages/memory/response_memory_probs.cpp (hunk 16 above)
Comment (dagardner-nv) [reply morpheus/_lib/src/messages/memory/response_memory_probs.cpp:40]: clang-format didn't like this change. I had to revert it.

Comment (mdemoret-nv) [reply morpheus/_lib/src/messages/memory/response_memory_probs.cpp:40]: Oh I see my mistake. Nevermind.

**Comment by mdemoret-nv** [comment]
@gpucibot merge

**Pull request merged**